UefiCpuPkg: Add Ro to Ap Buffer after copying ApLoopCode.

apop5 opened the pull request:
## Description

MpLib allocates a buffer for Ap Loop code. It manually removes the Nx protection but never applies any other protections to the buffer, leaving it RWX. After the code is copied into the allocated buffer, apply Read Only permissions so that nothing can overwrite it while the APs are executing from it.

In the update to 202502, Edk2 modified their code layout, resulting in the AP buffer being allocated with Nx, having Nx removed, and then never applying any other memory protections.


- [x] Impacts functionality?
- [x] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [x] Backport to release branch?

## How This Was Tested
Executing DxePagingAudit test application on Q35 found one region of memory marked RWX.
Applying these modifications, DxePagingAudit test application no longer reports a RWX buffer.

## Integration Instructions
No integration necessary.

apop5 commented:
@os-d 

Please review this, and if it is okay, I'll take it to EDK2 directly. Otherwise we can iterate on it here.

os-d commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -579,6 +579,34 @@ RemoveNxprotection (
+  EFI_STATUS                       Status;
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  MemDesc;
+
+  Status = gDS->GetMemorySpaceDescriptor (BaseAddress, &MemDesc);
```
We should check the capabilities and if RO is not set, set it.

os-d commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -579,6 +579,34 @@ RemoveNxprotection (
+  EFI_GCD_MEMORY_SPACE_DESCRIPTOR  MemDesc;
+
+  Status = gDS->GetMemorySpaceDescriptor (BaseAddress, &MemDesc);
+  if (!EFI_ERROR (Status)) {
```
Probably should assert if we hit an error in setting in any case.

os-d commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -601,12 +629,15 @@ MpInitChangeApLoopCallback (
   mNumberToFinish            = CpuMpData->CpuCount - 1;
-  // MU_CHANGE END Enable removal of NX attribute from buffer
-  BufferRemoveNoExecuteSetReadOnly (
+  RemoveNxprotection (
```
This is removing a Mu specific commit, right, but from a bigger commit, so probably we should do this as a separate commit (since can't take to edk2 that way) and mark that commit as SQUASH ON REBASE with the commit that added it

apop5 replied:
Agreed.
 
Planning on getting EDK2 PR up, give it a day or so, see if there is feedback.
Revert this commit
Revert the commit that introduced the BufferRemoveNoExecuteSetReadOnly,
Cherry-pick Edk2 (if its modified) then reapply the commit that brought BufferRemoveNoExecuteSetReadOnly

apop5 replied:
Revisiting this decision.

Looking at the commits that introduced Nx on the Ap Loop buffers:
https://github.com/microsoft/mu_basecore/commit/4ec393b76313973572f9cd1c1794a390192f2413

In the next integration, these will probably need to be dropped/reworked heavily to match with the Edk2 changes, and trying to unravel them now, while we know more changes are going into edk2, might not be worth the investment just to redo the work again.

Thoughts?

os-d replied:
That was my thought on the squash on rebase commit. We make it clear that it should be merged with the original commit in the next integration and then when trying to rebase that now unified commit in the next integration, we deal with what needs to change.

apop5 committed:
UefiCpuPkg: Apply ReadOnly on Ap loop buffers.

In the MpInitLib, pages are allocated for
the APs, and data is copied into them. Depending
on the allocation type, Nx needs to be removed
from the pages to allow the processors to execute.

Additionally, ReadOnly needs to be applied to the
pages after they have been filled with the
instructions that the APs need to execute.

Signed-off-by: Aaron Pop <aaronpop@microsoft.com>
Co-authored-by: Oliver Smith-Denny <osde@microsoft.com>

makubacki commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a - Removing Nx on 0x%p returned %r\n", __func__, BaseAddress, Status));
```
nit: Technically the error status code could be from getting the memory space descriptor not necessarily removing NX on the found descriptor.

makubacki commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
While you're in this code, could you please add a separate commit that makes `RemoveNxprotection()` match naming conventions? - `RemoveNxProtection()`

makubacki commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
+                        );
+      }
+    }
+  }
```
Is there a reason not to combine these? Both blocks lead to a single code path.

```suggestion
    if (((MemDesc.Capabilities & EFI_MEMORY_XP) == EFI_MEMORY_XP) && ((MemDesc.Attributes & EFI_MEMORY_XP) == EFI_MEMORY_XP)) {
      Status = gDS->SetMemorySpaceAttributes (
                      BaseAddress,
                      Length,
                      MemDesc.Attributes & (~EFI_MEMORY_XP)
                      );
    }
  }
```

If you find the existing way more readable, feel free to leave it. I find it a little more readable to have less nesting.

apop5 replied:
Implemented in the latest.

makubacki commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
+/**
+  Add ReadOnly protection to the range specified by BaseAddress and Length.
+
+  The PEI implementation uses CpuPageTableLib to change the attribute.
+  The DXE implementation uses gDS to change the attribute.
```
It is confusing to include the PEI implementation details here because this function is in `DxeMpLib.c`.

apop5 replied:
cleaned up in latest.

makubacki commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a - Setting Ro on 0x%p returned %r\n", __func__, BaseAddress, Status));
```
Similar comment to the other function about keeping the error message precise to where the error code originated.

makubacki commented on UefiCpuPkg/Library/MpInitLib/MpLib.h:
```diff
@@ -997,24 +997,19 @@ RemoveNxprotection (
-BufferRemoveNoExecuteSetReadOnly (
-  IN EFI_PHYSICAL_ADDRESS  Buffer,
-  IN UINTN                 Size
+VOID
+ApplyReadOnlyMemoryProtection (
```
```suggestion
VOID
EFIAPI
ApplyReadOnlyMemoryProtection (
```

apop5 replied:
Since this is consumed from inside of the PeiMpLib.inf and DxeMpLib.inf, do we need the EFIAPI? 
What I mean is that its not a external interface to the library, the functions are not defined in the 
UefiCpuPkg\Include\Library\MpInitLib.h

makubacki replied:
You don't. I saw MpLib.h and quickly categorized this as a public header at the time.

makubacki commented on UefiCpuPkg/Library/MpInitLib/PeiMpLib.c:
```diff
@@ -960,6 +936,25 @@ RemoveNxprotection (
+  @param[in] BaseAddress  BaseAddress of the range.
+  @param[in] Length       Length of the range.
+**/
+VOID
+ApplyReadOnlyMemoryProtection (
```
```suggestion
VOID
EFIAPI
ApplyReadOnlyMemoryProtection (
```

makubacki commented on UefiCpuPkg/Library/MpInitLib/DxeMpLib.c:
```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
+  @param[in] BaseAddress  BaseAddress of the range.
+  @param[in] Length       Length of the range.
+**/
+VOID
+ApplyReadOnlyMemoryProtection (
```
```suggestion
VOID
EFIAPI
ApplyReadOnlyMemoryProtection (
```

makubacki commented on UefiCpuPkg/Library/MpInitLib/PeiMpLib.c:
```diff
@@ -960,6 +936,25 @@ RemoveNxprotection (
+/**
+  Add ReadOnly protection to the range specified by BaseAddress and Length.
+
+  The PEI implementation uses CpuPageTableLib to change the attribute.
+  The DXE implementation uses gDS to change the attribute.
```
```suggestion
  The PEI implementation uses CpuPageTableLib to change the attribute.
```

apop5 committed:
Merge branch 'dev/202502' into personal/apop5/roonapbuffer
